File: languages/csharp/protobuf/Bench.cs
// Protobuf C# benchmark target.
// Encodes + decodes the canonical FullScaleExample message, hand-filled from
// schema/STATE.md, through protobuf's generated Fullscale types. Same message,
// same state, same timed region as the SofaBuffers target.
//
// ONE source, TWO impls, selected by the BENCH_IMPL env var (#108):
//
//   protobuf         ToByteArray() -- one buffer the size of the whole message
//   protobuf-stream  WriteTo(CodedOutputStream) over a bounded buffer flushed
//                    into a Stream as it fills -- protobuf's own answer to the
//                    question sofab-stream asks. It is the opponent the runner
//                    pairs with impl=sofab-stream.
using System;
using System.Diagnostics;
using System.IO;
using System.Security.Cryptography;
using Google.Protobuf;
using Fullscale;

static class Program {
    static FullScaleExample Build() {
        var m = new FullScaleExample {
            U8 = 200,
            I8 = -100,
            U16 = 50000,
            I16 = -20000,
            U32 = 3000000000,
            I32 = -1000000000,
            U64 = 10000000000000UL,
            I64 = -5000000000000L,
            Nested = new FullScaleSeqStruct {
                F32 = 3.14f,
                F64 = 3.14159265,
                Str = "Hello, World!",
                BytesField = ByteString.CopyFrom(new byte[] { 0xDE, 0xAD, 0xBE, 0xEF }),
            },
            Arrays = new FullScaleSeqStructOfArrays(),
            StringArray = new FullScaleSeqArrayOfStrings(),
        };
        var a = m.Arrays;
        a.U8.AddRange(new uint[] { 0, 64, 128, 191, 255 });
        a.I8.AddRange(new int[] { -128, -64, 0, 63, 127 });
        a.U16.AddRange(new uint[] { 0, 16384, 32768, 49151, 65535 });
        a.I16.AddRange(new int[] { -32768, -16384, 0, 16383, 32767 });
        a.U32.AddRange(new uint[] { 0, 1073741824, 2147483648, 3221225471, 4294967295 });
        a.I32.AddRange(new int[] { -2147483648, -1073741824, 0, 1073741823, 2147483647 });
        a.U64.AddRange(new ulong[] {
            0, 4611686018427387904, 9223372036854775808,
            13835058055282163711, 18446744073709551615 });
        a.I64.AddRange(new long[] {
            -9223372036854775807, -4611686018427387904, 0,
            4611686018427387903, 9223372036854775807 });
        a.Nested = new FullScaleSeqStructOfFpArrays();
        a.Nested.Fp32.AddRange(new float[] { 1f, 2f, 3f, -float.MaxValue, float.MaxValue });
        a.Nested.Fp64.AddRange(new double[] { 1d, 2d, 3d, -double.MaxValue, double.MaxValue });
        m.StringArray.Strings.AddRange(new string[] {
            "Hello, Sofab!", "", "1234567890", "äöüÄÖÜß",
            "This_is_a_very_long_test_string_with_!@#$%^&*()_+-=[]{}" });
        return m;
    }

    // Same sink shape as the sofab harness: the drained bytes land in a fixed
    // array, so the row compares codecs and not the harness around them.
    static readonly byte[] Sink = new byte[2048];
    static int sinkN;

    sealed class SinkStream : Stream {
        public override bool CanRead => false;
        public override bool CanSeek => false;
        public override bool CanWrite => true;
        public override long Length => sinkN;
        public override long Position { get => sinkN; set => throw new NotSupportedException(); }
        public override void Flush() { }
        public override int Read(byte[] b, int o, int c) => throw new NotSupportedException();
        public override long Seek(long o, SeekOrigin r) => throw new NotSupportedException();
        public override void SetLength(long v) => throw new NotSupportedException();
        public override void Write(byte[] b, int o, int c) {
            Buffer.BlockCopy(b, o, Sink, sinkN, c);
            sinkN += c;
        }
    }

    static readonly SinkStream SinkOut = new SinkStream();

    /// <summary>Streaming encode through a <paramref name="cap"/>-byte buffer; returns bytes drained.</summary>
    static int StreamEncode(FullScaleExample m, int cap) {
        sinkN = 0;
        var cos = new CodedOutputStream(SinkOut, cap, leaveOpen: true);
        m.WriteTo(cos);
        cos.Flush();
        return sinkN;
    }

    static int Main() {
        var src = Build();
        var parser = FullScaleExample.Parser;

        string impl = Environment.GetEnvironmentVariable("BENCH_IMPL") ?? "protobuf";
        bool streaming = impl == "protobuf-stream";
        int streamCap = int.Parse(Environment.GetEnvironmentVariable("STREAM_BUF_BYTES") ?? "64");

        // Warm-up round-trip + self-check (outside the timed region).
        byte[] blob = src.ToByteArray();
        int serialized = blob.Length;
        string sha = Convert.ToHexString(SHA256.HashData(blob)).ToLowerInvariant();
        // The check runs through whichever encode path this impl measures.
        bool ok = streaming
            ? StreamEncode(parser.ParseFrom(blob), streamCap) == serialized
              && ((ReadOnlySpan<byte>)Sink).Slice(0, serialized).SequenceEqual(blob)
            : ((ReadOnlySpan<byte>)parser.ParseFrom(blob).ToByteArray()).SequenceEqual(blob);
        if (!ok) {
            Console.Error.WriteLine("FAIL: protobuf round-trip self-check");
            Environment.Exit(1);
        }

        long iters = long.Parse(Environment.GetEnvironmentVariable("BENCH_ITERS") ?? "2000000");

        // JIT warm-up (same chained shape as the timed loop).
        for (int i = 0; i < 5000; i++) {
            if (streaming) StreamEncode(parser.ParseFrom(blob), streamCap);
            else parser.ParseFrom(blob).ToByteArray();
        }

        // Chained round trip: decode the reference wire, then re-encode the freshly
        // parsed message (issue #86) — the proxy/transcode shape. Each ParseFrom
        // yields a new message whose cached serialized size is unset, so protobuf
        // pays the size pass every encode instead of hitting a once-per-instance
        // memo. sink keeps the re-encode live and doubles as a loop-path check.
        //
        // Two loops, picked before the stopwatch starts, so neither impl pays a
        // per-iteration branch the other does not.
        long sink = 0;
        Stopwatch sw;
        if (streaming) {
            sw = Stopwatch.StartNew();
            for (long i = 0; i < iters; i++) {
                sink += StreamEncode(parser.ParseFrom(blob), streamCap);
            }
        } else {
            sw = Stopwatch.StartNew();
            for (long i = 0; i < iters; i++) {
                sink += parser.ParseFrom(blob).ToByteArray().Length;
            }
        }
        sw.Stop();

        if (sink != (long)serialized * iters
            || (streaming && !((ReadOnlySpan<byte>)Sink).Slice(0, serialized).SequenceEqual(blob))) {
            Console.Error.WriteLine("FAIL: protobuf loop-path self-check");
            Environment.Exit(1);
        }

        double cpu = sw.Elapsed.TotalSeconds;
        double mbs = cpu > 0 ? (double)serialized * iters / cpu / 1e6 : 0.0;
        Console.WriteLine(
            $"BENCH lang=csharp impl={impl} serialized_bytes={serialized} iters={iters} " +
            $"cpu_time_s={cpu:F6} throughput_mbs={mbs:F2} sha256={sha}");
        return 0;
    }
}

File: languages/cpp/protobuf/bench.cpp
// Protobuf C++ benchmark target.
//
// Encodes + decodes the canonical FullScaleExample message (schema/STATE.md)
// through protobuf's generated message.pb types. Same message, same state, same
// timed region as the SofaBuffers target. Prints one uniform BENCH line.
//
// ONE source, TWO builds, mirroring sofab/bench.cpp:
//
//   impl=protobuf         SerializeToString — one growable buffer per encode
//   impl=protobuf-stream  #108: BENCH_STREAM — SerializeToZeroCopyStream over a
//                         bounded block drained as it fills, protobuf's own answer
//                         to the same question. It is the opponent the runner
//                         pairs with impl=sofab-stream.
#ifndef _POSIX_C_SOURCE
#  define _POSIX_C_SOURCE 199309L
#endif

// Which impl this build measures; set by setup.sh (-DBENCH_IMPL).
#ifndef BENCH_IMPL
#  define BENCH_IMPL "protobuf"
#endif

#include <cfloat>
#include <cstddef>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <ctime>
#include <string>

#ifdef BENCH_STREAM
#  include <cstring>
#  include <google/protobuf/io/zero_copy_stream.h>
#endif

#include "message.pb.h"
#include "sha256.h"

namespace {

#ifdef BENCH_STREAM
// The streaming counterpart of sofab's OStreamView-with-flush-callback: a
// ZeroCopyOutputStream that hands protobuf a bounded block and drains it into the
// sink each time protobuf asks for the next one. Same buffer size and same sink
// shape as sofab/bench.cpp, so the row differs in the codec and not in the
// harness around it.
#  ifndef STREAM_BUF_BYTES
#    define STREAM_BUF_BYTES 64
#  endif

std::uint8_t g_sink[2048];
std::size_t  g_sink_n = 0;

class DrainingStream final : public google::protobuf::io::ZeroCopyOutputStream
{
public:
    // protobuf takes back the bytes it did not use via BackUp(), so the amount
    // actually drained is `handed out - backed up` — tracked in pending_.
    bool Next(void **data, int *size) override
    {
        drain();
        *data = block_;
        *size = static_cast<int>(sizeof block_);
        pending_ = sizeof block_;
        return true;
    }
    void BackUp(int count) override { pending_ -= static_cast<std::size_t>(count); }
    google::protobuf::int64 ByteCount() const override
    {
        return static_cast<google::protobuf::int64>(total_ + pending_);
    }
    // Hand over whatever is still in the block; call once the encode is done.
    void finish() { drain(); }

private:
    void drain()
    {
        if (pending_ != 0) {
            std::memcpy(g_sink + g_sink_n, block_, pending_);
            g_sink_n += pending_;
            total_   += pending_;
            pending_  = 0;
        }
    }
    std::uint8_t block_[STREAM_BUF_BYTES];
    std::size_t  pending_ = 0;
    std::size_t  total_   = 0;
};

std::size_t stream_encode(const fullscale::FullScaleExample &m)
{
    g_sink_n = 0;
    DrainingStream out;
    m.SerializeToZeroCopyStream(&out);
    out.finish();
    return g_sink_n;
}
#endif

double now_seconds()
{
    struct timespec ts;
    clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &ts);
    return static_cast<double>(ts.tv_sec) + static_cast<double>(ts.tv_nsec) / 1.0e9;
}

long bench_iters(long fallback)
{
    const char *s = getenv("BENCH_ITERS");
    if (s != nullptr && s[0] != '\0') {
        long v = atol(s);
        if (v > 0) {
            return v;
        }
    }
    return fallback;
}

} // namespace

int main()
{
    GOOGLE_PROTOBUF_VERIFY_VERSION;

    fullscale::FullScaleExample msg;

    msg.set_u8(200);
    msg.set_i8(-100);
    msg.set_u16(50000);
    msg.set_i16(-20000);
    msg.set_u32(3000000000U);
    msg.set_i32(-1000000000);
    msg.set_u64(10000000000000ULL);
    msg.set_i64(-5000000000000LL);

    fullscale::FullScaleSeqStruct* nested = msg.mutable_nested();
    nested->set_f32(3.14f);
    nested->set_f64(3.14159265);
    nested->set_str("Hello, World!");
    nested->set_bytes_field(std::string("\xDE\xAD\xBE\xEF", 4));

    fullscale::FullScaleSeqStructOfArrays* arrays = msg.mutable_arrays();
    for (uint32_t v : {0U, 64U, 128U, 191U, 255U}) arrays->add_u8(v);
    for (int32_t v : {-128, -64, 0, 63, 127}) arrays->add_i8(v);
    for (uint32_t v : {0U, 16384U, 32768U, 49151U, 65535U}) arrays->add_u16(v);
    for (int32_t v : {-32768, -16384, 0, 16383, 32767}) arrays->add_i16(v);
    for (uint32_t v : {0U, 1073741824U, 2147483648U, 3221225471U, 4294967295U}) arrays->add_u32(v);
    for (int32_t v : {INT32_MIN, -1073741824, 0, 1073741823, INT32_MAX}) arrays->add_i32(v);
    for (uint64_t v : {0ULL, 4611686018427387904ULL, 9223372036854775808ULL,
                       13835058055282163711ULL, 18446744073709551615ULL}) arrays->add_u64(v);
    for (int64_t v : {-9223372036854775807LL, -4611686018427387904LL, 0LL,
                      4611686018427387903LL, 9223372036854775807LL}) arrays->add_i64(v);

    fullscale::FullScaleSeqStructOfFpArrays* na = arrays->mutable_nested();
    for (float v : {1.0f, 2.0f, 3.0f, -FLT_MAX, FLT_MAX}) na->add_fp32(v);
    for (double v : {1.0, 2.0, 3.0, -DBL_MAX, DBL_MAX}) na->add_fp64(v);

    fullscale::FullScaleSeqArrayOfStrings* sa = msg.mutable_string_array();
    sa->add_strings("Hello, Sofab!");
    sa->add_strings("");
    sa->add_strings("1234567890");
    sa->add_strings("\xC3\xA4\xC3\xB6\xC3\xBC\xC3\x84\xC3\x96\xC3\x9C\xC3\x9F"); // "äöüÄÖÜß"
    sa->add_strings("This_is_a_very_long_test_string_with_!@#$%^&*()_+-=[]{}");

    // Warm-up round-trip + self-check (outside the timed region).
    std::string buffer;
    if (!msg.SerializeToString(&buffer)) {
        fprintf(stderr, "FAIL: protobuf serialize\n");
        return 1;
    }
    const size_t serialized = buffer.size();
    char sha[65];
    sha256_hex(buffer.data(), serialized, sha);

    fullscale::FullScaleExample check;
    if (!check.ParseFromString(buffer)) {
        fprintf(stderr, "FAIL: protobuf parse\n");
        return 1;
    }
#ifdef BENCH_STREAM
    if (stream_encode(check) != serialized ||
        std::memcmp(g_sink, buffer.data(), serialized) != 0) {
        fprintf(stderr, "FAIL: protobuf round-trip self-check\n");
        return 1;
    }
#else
    std::string reencoded;
    if (!check.SerializeToString(&reencoded) || reencoded != buffer) {
        fprintf(stderr, "FAIL: protobuf round-trip self-check\n");
        return 1;
    }
#endif

    // Timed loop: chained round trip — decode the reference wire, then re-encode
    // the freshly parsed message (issue #86). ParseFromString clears `dec` first,
    // so each iteration re-parses into a message whose cached serialized size is
    // reset — protobuf therefore pays the size pass every encode, as a real
    // proxy/transcode would, instead of hitting a once-per-instance memo.
    //
    // The DECODE half stays the plain one-shot parse in both builds — this row
    // varies the encode path only (#108).
    fullscale::FullScaleExample dec;
#ifndef BENCH_STREAM
    std::string out;
#endif
    const long iters = bench_iters(500000);
    const double t0 = now_seconds();
    for (long i = 0; i < iters; ++i) {
        dec.ParseFromString(buffer);
#ifdef BENCH_STREAM
        stream_encode(dec);
#else
        out.clear();
        dec.SerializeToString(&out);
#endif
    }
    const double t1 = now_seconds();

#ifdef BENCH_STREAM
    if (g_sink_n != serialized ||
        std::memcmp(g_sink, buffer.data(), serialized) != 0 ||
#else
    if (out != buffer ||
#endif
        dec.u8() != msg.u8() ||
        dec.string_array().strings_size() != 5 ||
        dec.arrays().i64_size() != 5) {
        fprintf(stderr, "FAIL: protobuf loop-path self-check\n");
        return 1;
    }

    const double cpu = t1 - t0;
    const double mbs = (cpu > 0.0)
        ? static_cast<double>(serialized) * static_cast<double>(iters) / cpu / 1.0e6
        : 0.0;
    printf("BENCH lang=cpp impl=%s serialized_bytes=%zu iters=%ld "
           "cpu_time_s=%.6f throughput_mbs=%.2f sha256=%s\n",
           BENCH_IMPL, serialized, iters, cpu, mbs, sha);
    return 0;
}
